### PR TITLE
s2i: build hung, stuck

### PR DESCRIPTION
**Pull request by gabemontero**
Bug 1370265
Bugzilla link https://bugzilla.redhat.com/show_bug.cgi?id=1370265
Detailed explanation
a timeout on container start could result on waitgroup.Done being called twice, the waitgroup going negative, and down a hung/panic path

the refactoring of pull 576 completed with this pull

@bparees @csrwng @rhcarvalho PTAL

### CODE REVIEW

**Comment by gabemontero** [comment]
reposting comment from @bparees in pr 576 for convenience:

@rhcarvalho would really appreciate if you could take a look at this fix, we're trying to resolve a blocker bug for 3.3. The go stacks for the hang we're trying to address are in the referenced bug.

The net is wg.Done() could get called twice (once for the source upload and once for the timeout error on starting the container)

**Comment by bparees** [comment]
@rhcarvalho belated apologies for tagging you in all this stuff on your PTO, forgot you were out.  So..welcome back and ptal when you can :)

**Comment by gabemontero** [comment]
+1 @rhcarvalho :-)

**Comment by gabemontero** [comment]
@rhcarvalho do you think you'll be able to take a look at this?  We are interested in your perspective as the one who originally introduced (if I read `git blame` correctly) the use of the wait group in this path.

thanks

**Comment by rhcarvalho** [comment]
Sorry for the delay. Looking at it.

**Comment by gabemontero** [comment]
no worries @rhcarvalho - we know you are super busy with feed henry

**Comment by gabemontero** [comment]
thanks for looking

File: pkg/build/strategies/sti/sti.go [hunk 1]
```diff
@@ -596,9 +595,8 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
 	go dockerpkg.StreamContainerIO(errReader, &errOutput, func(a ...interface{}) { glog.Info(a...) })
 
 	err := builder.docker.RunContainer(opts)
-	if util.IsTimeoutError(err) {
-		// Cancel waiting for source input if the container timeouts
-		wg.Done()
+	if !config.LayeredBuild && !util.IsTimeoutError(err) {
+		defer wg.Wait()
```
Comment (rhcarvalho) [inline pkg/build/strategies/sti/sti.go]: @gabemontero @bparees good job. Reading this method is an adventure... it deserves being rewritten, but not right now, let's get the issue fixed first.

Comment (gabemontero) [inline pkg/build/strategies/sti/sti.go]: cool - thx @rhcarvalho

**Comment by gabemontero** [comment]
closing in favor of https://github.com/openshift/source-to-image/issues/581

**Pull request closed without merging**